File: builder-pattern/python/src/Application.py
# -*- coding: utf-8 -*-
"""
@author: jarry
"""
import json
from src.Director import Director
from src.PhoneBuilder import PhoneBuilder
from src.ManualBuilder import ManualBuilder


class Application():

    def __init__(self):
        pass

    @staticmethod
    def makeIPhone():
        director = Director()
        # 创建手机
        phoneBuilder = PhoneBuilder()
        director.buildIPhone(phoneBuilder)
        iPhone = phoneBuilder.getProduct()
        print('iPhone:' + iPhone.getName() + ' | ', json.dumps(iPhone.__dict__))

        # 创建手册
        manualBuilder = ManualBuilder()
        director.buildIPhone(manualBuilder)
        manual = manualBuilder.getProduct()
        print('manual:' + manual.getName() + ' | ', json.dumps(manual.__dict__))

    @staticmethod
    def makeHuaweiPhone():
        director = Director()

        # 创建手机
        phoneBuilder = PhoneBuilder()
        director.buildHuaweiPhone(phoneBuilder)
        huaweiPhone = phoneBuilder.getProduct()
        print('huaweiPhone:' + huaweiPhone.getName() + ' | ', json.dumps(huaweiPhone.__dict__))

        # 创建手册
        manualBuilder = ManualBuilder()
        director.buildHuaweiPhone(manualBuilder)
        manual = manualBuilder.getProduct()
        print('manual:' + manual.getName() + ' | ', json.dumps(manual.__dict__))

File: builder-pattern/ts/src/Application.ts
import { Director } from './Director.js'
import { PhoneBuilder } from './PhoneBuilder.js'
import { ManualBuilder } from './ManualBuilder.js'

export class Application {
  static makeIPhone() {
    const director = new Director()
    // 创建手机
    const phoneBuilder = new PhoneBuilder()
    director.buildIPhone(phoneBuilder)
    const iPhone = phoneBuilder.getProduct()
    console.log('iPhone:' + iPhone.getName() + ' | ', JSON.stringify(iPhone))

    // 创建手册
    const manualBuilder = new ManualBuilder()
    director.buildIPhone(manualBuilder)
    const manual = manualBuilder.getProduct()
    console.log('manual:' + manual.getName() + ' | ', JSON.stringify(manual))
  }

  static makeHuaweiPhone() {
    const director = new Director()

    // 创建手机
    const phoneBuilder = new PhoneBuilder()
    director.buildHuaweiPhone(phoneBuilder)
    const huaweiPhone = phoneBuilder.getProduct()
    console.log(
      'huaweiPhone:' + huaweiPhone.getName() + ' | ',
      JSON.stringify(huaweiPhone)
    )

    // 创建手册
    const manualBuilder = new ManualBuilder()
    director.buildHuaweiPhone(manualBuilder)
    const manual = manualBuilder.getProduct()
    console.log('manual:' + manual.getName() + ' | ', JSON.stringify(manual))
  }
}
